### PR TITLE
Add unit tests for service classes

### PR DESCRIPTION
**Pull request by majabrankovic**
Added unit tests for UserService, RoleService, ProfileService, PostService and UserRoleService
Covered basic CRUD operations and validation scenarios

### CODE REVIEW

File: src/test/java/com/example/demo/service/PostServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    private PostRepository postRepository;
+
+    @InjectMocks
+    private PostService postService;
+
+    private Post post;
+
+    @BeforeEach
+    public void setUp() {
+        post = Post.builder()
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/PostServiceTest.java:33]: Is this object common for all tests?

File: src/test/java/com/example/demo/service/PostServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                .id(1L)
+                .title("Naslov")
+                .content("Sadrzaj posta")
+                .build();
+
+    }
+
+    @Test
+    public void PostService_CreatePost_ReturnPost() {
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/PostServiceTest.java:42]: Add test for case that you pass null to `createPost()` mehod

File: src/test/java/com/example/demo/service/PostServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+    @Test
+    public void PostService_GetPostById_ReturnEmptyPostWhenPostDoesNotExist() {
+        when(postRepository.findById(99L)).thenReturn(Optional.empty());
+
+        Post result = postService.getPostById(99L);
+
+        Assertions.assertThat(result).isNotNull();
+        Assertions.assertThat(result.getId()).isNull();
+    }
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/PostServiceTest.java:68]: Add test for get null value for ID

File: src/test/java/com/example/demo/service/PostServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    @Test
+    public void PostService_UpdatePost_ReturnEmptyPostWhenPostDoesNotExist() {
+        when(postRepository.findById(post.getId())).thenReturn(Optional.empty());
+
+        Post result = postService.updatePost(post);
+
+        Assertions.assertThat(result.getId()).isNull();
+    }
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/PostServiceTest.java:97]: Add test for update null post

File: src/test/java/com/example/demo/service/PostServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    @Test
+    public void PostService_DeletePostById_ReturnPostDeleted() {
+        Long postId = 1L;
+
+        postService.deletePostById(postId);
+
+        verify(postRepository, times(1)).deleteById(postId);
+    }
+
+}
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/PostServiceTest.java:108]: Add tests for deleting post with invalid and null ID

**Review by vladanskoko — CHANGES_REQUESTED** [review]

File: src/test/java/com/example/demo/service/PostServiceTest.java (hunk 1 above)
Comment (majabrankovic) [reply src/test/java/com/example/demo/service/PostServiceTest.java:33]: The post object is initialized before each test, so it's a fresh instance for every test.

File: src/test/java/com/example/demo/service/PostServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+import java.util.Optional;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+public class PostServiceTest {
+
+    @Mock
+    private PostRepository postRepository;
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/PostServiceTest.java:24]: All mocks name with suffix `Mock` in this case `postRepositoryMock`

File: src/test/java/com/example/demo/service/PostServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+    @InjectMocks
+    private PostService postService;
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/PostServiceTest.java:27]: Why service is mock?

File: src/test/java/com/example/demo/service/PostServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    @Test
+    public void PostService_CreatePost_ReturnPost() {
+        when(postRepository.save(post)).thenReturn(post);
+
+        Post result = postService.createPost(post);
+
+        Assertions.assertThat(result).isEqualTo(post);
+        verify(postRepository).save(post);
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/PostServiceTest.java:48]: Verify how many times mock executed

File: src/test/java/com/example/demo/service/PostServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    @Test
+    public void PostService_GetPostById_ReturnPostWhenPostExists() {
+        when(postRepository.findById(1L)).thenReturn(Optional.of(post));
+
+        Post result = postService.getPostById(1L);
+
+        Assertions.assertThat(result).isEqualTo(post);
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/PostServiceTest.java:57]: Add `verify()`

File: src/test/java/com/example/demo/service/PostServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    @Test
+    public void PostService_GetPostById_ReturnEmptyPostWhenPostDoesNotExist() {
+        when(postRepository.findById(99L)).thenReturn(Optional.empty());
+
+        Post result = postService.getPostById(99L);
+
+        Assertions.assertThat(result).isNotNull();
+        Assertions.assertThat(result.getId()).isNull();
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/PostServiceTest.java:67]: Add `verify()`

File: src/test/java/com/example/demo/service/PostServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    @Test
+    public void PostService_GetAllPosts_ReturnPosts() {
+        List<Post> postList = List.of(post);
+        when(postRepository.findAll()).thenReturn(postList);
+
+        List<Post> result = postService.getAllPosts();
+
+        Assertions.assertThat(result).isEqualTo(postList);
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/PostServiceTest.java:77]: add `verify()` to all tests

**Review by vladanskoko — CHANGES_REQUESTED** [review]

File: src/test/java/com/example/demo/service/ProfileServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+import java.util.Optional;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+public class ProfileServiceTest {
+
+    @Mock
+    private ProfileRepository profileRepository;
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/ProfileServiceTest.java:24]: Change name to `profileRepositoryMock`

File: src/test/java/com/example/demo/service/ProfileServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+    @Test
+    public void ProfileService_CreateProfile_ReturnProfile() {
+        when(profileRepository.save(profile)).thenReturn(profile);
+
+        Profile result = profileService.createProfile(profile);
+
+        Assertions.assertThat(result).isEqualTo(profile);
+        verify(profileRepository).save(profile);
+    }
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/ProfileServiceTest.java:49]: Add test for creating null profile

File: src/test/java/com/example/demo/service/PostServiceTest.java (hunk 8 above)
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/PostServiceTest.java:48]: Add test for creating null post

File: src/test/java/com/example/demo/service/ProfileServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+    @Test
+    public void ProfileService_GetProfileById_ReturnEmptyProfileWhenProfileDoesNotExist() {
+        when(profileRepository.findById(99L)).thenReturn(Optional.empty());
+
+        Profile result = profileService.getProfileById(99L);
+
+        Assertions.assertThat(result).isNotNull();
+        Assertions.assertThat(result.getId()).isNull();
+    }
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/ProfileServiceTest.java:68]: Add test for returning null profile

File: src/test/java/com/example/demo/service/ProfileServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+    @Test
+    public void ProfileService_UpdateProfile_ReturnEmptyProfileWhenProfileDoesNotExist() {
+        when(profileRepository.findById(profile.getId())).thenReturn(Optional.empty());
+
+        Profile result = profileService.updateProfile(profile);
+
+        Assertions.assertThat(result.getId()).isNull();
+    }
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/ProfileServiceTest.java:97]: Add test for update null profile

File: src/test/java/com/example/demo/service/ProfileServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+
+    @Test
+    public void ProfileService_DeleteProfileById_ReturnProfileDeleted() {
+        Long profileId = 1L;
+
+        profileService.deleteProfileById(profileId);
+
+        verify(profileRepository, times(1)).deleteById(profileId);
+    }
+
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/ProfileServiceTest.java:107]: Add test for deleting profile with invalid ID and null ID

File: src/test/java/com/example/demo/service/ProfileServiceTest.java [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+        verify(profileRepository).save(profile);
+    }
+
+    @Test
+    public void ProfileService_GetProfileById_ReturnProfileWhenProfileExists() {
+        when(profileRepository.findById(1L)).thenReturn(Optional.of(profile));
+
+        Profile result = profileService.getProfileById(1L);
+
+        Assertions.assertThat(result).isEqualTo(profile);
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/ProfileServiceTest.java:57]: verify how many times mocks are invoked in all tests

File: src/test/java/com/example/demo/service/RoleServiceTest.java [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+import java.util.Optional;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+public class RoleServiceTest {
+
+    @Mock
+    private RoleRepository roleRepository;
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/RoleServiceTest.java:24]: Change name to `roleRepositoryMock`

File: src/test/java/com/example/demo/service/RoleServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+
+    @BeforeEach
+    void setUp() {
+        role = Role.builder()
+                .id(1L)
+                .name("ADMIN")
+                .build();
+    }
+
+    @Test
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/RoleServiceTest.java:39]: The same comments as for previous tests

File: src/test/java/com/example/demo/service/UserRoleServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+import java.util.Optional;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+public class UserRoleServiceTest {
+
+    @Mock
+    UserRoleRepository userRoleRepository;
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/UserRoleServiceTest.java:26]: The same comments as for previous tests

File: src/test/java/com/example/demo/service/UserServiceTest.java [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+import java.util.Optional;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+public class UserServiceTest {
+
+    @Mock
+    private UserRepository userRepository;
```
Comment (vladanskoko) [inline src/test/java/com/example/demo/service/UserServiceTest.java:23]: The same as for previous tests

**Review by vladanskoko — CHANGES_REQUESTED** [review]
